hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java

```diff
@@ -271,17 +272,22 @@ private void validateClusterVersionChange(Version newClusterVersion) {
         }
     }
 
-    private void checkMigrationsAndPartitionStateVersion(ClusterStateChange stateChange, int partitionStateVersion) {
-        final InternalPartitionService partitionService = node.getPartitionService();
-        final int thisPartitionStateVersion = partitionService.getPartitionStateVersion();
+    private void checkMigrationsAndPartitionStateStamp(ClusterStateChange stateChange, long partitionStateStamp) {
+        InternalPartitionService partitionService = node.getPartitionService();
+        long thisPartitionStateStamp;
+        if (clusterVersion.isGreaterOrEqual(Versions.V4_1)) {
+            thisPartitionStateStamp = partitionService.getPartitionStateStamp();
+        } else {
+            thisPartitionStateStamp = partitionService.getPartitionStateVersion();
```

Comment:
can't we remove `partitionStateStamp` from code-base and continue to use `partitionStateVersion` instead? It also helps hot-restart partial restart case.

Reply:
Problem is `partitionStateVersion` indicates a monotonic order between different partition states. But with the new mechanism, there's no such an order, members may receive partition updates in different order but still can have the same version. `partitionStateStamp` tries to solve this problem.